presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -0,0 +1,192 @@
+
+    public DecimalAverageAggregation()
+    {
+        super(NAME, ImmutableList.of(Signature.withVariadicBound("T", DECIMAL)), "T", ImmutableList.of("T"));
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Calculates the average value";
+    }
+
+    @Override
+    public InternalAggregationFunction specialize(Map<String, Type> types, List<TypeSignature> parameterTypes, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type type = types.get("T");
+        return generateAggregation(type);
+    }
+
+    private static InternalAggregationFunction generateAggregation(Type type)
+    {
+        DynamicClassLoader classLoader = new DynamicClassLoader(DecimalAverageAggregation.class.getClassLoader());
+
+        List<Type> inputTypes = ImmutableList.of(type);
+
+        MethodHandle inputFunction;
+        MethodHandle combineFunction;
+        MethodHandle outputFunction = null;
+        Class<? extends AccumulatorState> stateInterface = BigIntegerAndLongState.class;
+        AccumulatorStateSerializer<?> stateSerializer = new BigIntegerAndLongStateSerializer();
+
+        if (type.getJavaType() == long.class) {
+            inputFunction = SHORT_DECIMAL_INPUT_FUNCTION;
+            outputFunction = SHORT_DECIMAL_OUTPUT_FUNCTION;
+            combineFunction = COMBINE_FUNCTION;
+        }
+        else {
+            inputFunction = LONG_DECIMAL_INPUT_FUNCTION;
+            outputFunction = LONG_DECIMAL_OUTPUT_FUNCTION;
+            combineFunction = COMBINE_FUNCTION;
+        }
+        inputFunction = inputFunction.bindTo(type);
+        outputFunction = outputFunction.bindTo(type);
+
+        AggregationMetadata metadata = new AggregationMetadata(
+                generateAggregationName(NAME, type, inputTypes),
+                createInputParameterMetadata(type),
+                inputFunction,
+                null,
+                null,
+                combineFunction,
+                outputFunction,
+                stateInterface,
+                stateSerializer,
+                new BigIntegerAndLongStateFactory(),
+                type,
+                false);
+
+        Type intermediateType = stateSerializer.getSerializedType();
+        GenericAccumulatorFactoryBinder factory = new AccumulatorCompiler().generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(NAME, inputTypes, intermediateType, type, true, false, factory);
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type type)
+    {
+        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(BLOCK_INPUT_CHANNEL, type), new ParameterMetadata(BLOCK_INDEX));
+    }
+
+    public static void inputShortDecimal(Type type, BigIntegerAndLongState state, Block block, int position)
+    {
+        initializeIfNeeded(state);
```

Comment:
my first thought was to move this initialisation to `BigIntegerAndLongStateFactory` where the type is created, but the other implementation do it here so I'm following this pattern, although I personally believe it's the factory's responsibility.